website/docs/docs/building-a-dbt-project/building-models/python-models.md

```diff
@@ -0,0 +1,378 @@
+
+Python models have limited access to project context, including inputs such as `var`, `env_var`, and `target`. If you want to use those values to power conditional logic in your model, we recommend passing them through a dedicated `.yml` file config instead:
+
+<File name='models/config.yml'>
+
+```yml
+version: 2
+
+models:
+  - name: my_python_model
+    config:
+      materialized: table
+      target_name: "{{ target.name }}"
+      specific_env_var: "{{ env_var('SPECIFIC_ENV_VAR') }}"
+```
+
+</File>
+
+Then, use the `dbt.config.get()` function to _access_ values of configurations that have been set:
+
+<File name='models/my_python_model.py'>
+
+```python
+def model(dbt, session):
+    ...
+    target_name = dbt.config.get("target_name")
+    orders_df = dbt.ref("fct_orders")
+    
+    # limit data in dev
+    if target_name == "dev":
+        orders_df = orders_df.limit(500)
+```
+
+</File>
+
+### Materializations
+
+Python models support two materializations:
+- `table`
+- `incremental`
+
+For incremental models, similar to SQL models, you will need to filter incoming tables to just new rows of data:
+
+<WHCode>
+
+<div warehouse="Snowpark">
+
+<File name='models/my_python_model.py'>
+
+```python
+import snowflake.snowpark.functions as F
+
+def model(dbt, session):
+    dbt.config(materialized = "incremental")
+    df = dbt.ref("upstream_table")
+    
+    if dbt.is_incremental:
+
+        # only new rows compared to max in current table
+        max_from_this = f"select max(updated_at) from {dbt.this}"
+        df = df.filter(df.updated_at >= session.sql(max_from_this).collect()[0][0])
+
+        # or only rows from the past 3 days
+        df = df.filter(df.updated_at >= F.dateadd("day", F.lit(-3), F.current_timestamp()))
+        
+    ...
+        
+    return df
+```
+
+</File>
+
+</div>
+
+<div warehouse="PySpark">
+
+<File name='models/my_python_model.py'>
+
+```python
+import pyspark.sql.functions as F
+
+def model(dbt, session):
+    dbt.config(materialized = "incremental")
+    df = dbt.ref("upstream_table")
+    
+    if dbt.is_incremental:
+
+        # only new rows compared to max in current table
+        max_from_this = f"select max(updated_at) from {dbt.this}"
+        df = df.filter(df.updated_at >= session.sql(max_from_this).collect()[0][0])
+
+        # or only rows from the past 3 days
+        df = df.filter(df.updated_at >= F.date_add(F.current_timestamp(), F.lit(-3)))
+    
+    ...
+    
+    return df
+```
+
+</File>
+
+</div>
+
+</WHCode>
+
+**Note:** Incremental models are not yet supported on BigQuery.
+
+## Python-specific functionality
+
+### Defining functions
+
+In addition to defining a `model` function, the Python model can import other functions, or define its own functions. Here's an example, on Snowpark, defining a custom `add_one` function:
+
+<File name='models/my_python_model.py'>
+
+```python
+def add_one(x):
+    return x + 1
```

Comment:
Currently you cannot do that. It is possible to support(not very pretty way). @jtcohen6 maybe that's something to call out that we don't support it.

Reply:
I really consider this an open question for us!

I'm disinclined to do this one, given the chaos it could lead to, and over-reliance on the local file system:
> Or can you do something chaotic like `import add_one from my_python_model` in `my_second_python_model.py`?

I think we have two good options to pursue:
- Some platforms support registering these functions as "named" / persistent UDFs. This would be the case for "dbt should know about `function`s." These UDFs have some drawbacks; inconsistent support; and the main use case seems to be defining a Python UDF that then gets called from within a SQL model. That seems valuable for exposing a prediction function output of a trained ML model, less so for a utility like `add_one`.
- Potentially more promising: most of these platforms allow you to upload your _own_ home-spun "Python package," as a `.zip`, ==. You could then `import` it and use its functions, as if it were a public PyPI package.

Reply:
I agree that you shouldn't be able to do arbitrary imports from functions defined in other models in the `models/` directory.

> most of these platforms allow you to upload your own home-spun "Python package," as a .zip, ==. You could then import it and use its functions, as if it were a public PyPI package

Where would the code that gets zipped and uploaded be source controlled? This feels very analogous to the before-state of https://discourse.getdbt.com/t/using-dbt-to-manage-user-defined-functions/18 - code is magically available but no one really knows where it came from. 

If we can resolve that, I do like the idea of an internal functions package. 

Would finding `.py` functions defined in the `macros` directory (or add `functions/` to your `macro-paths`) and embedding them all into an internal package be good enough at breaking reliance on the filesystem? This assumes that you're disinclined to enable imports from inside a model because it would be bad practice, not that it would be hard to implement technically.

Reply:
> Where would the code that gets zipped and uploaded be source controlled?

Yup — 100% needs to live within the dbt project, and be version-controlled along with it.

I think that "upload" step will require different behind-the-scenes implementations across different data warehouses. We could even opt for UDFs as the backing implementation, if a given warehouse doesn't support arbitrary package upload. There's risk of a leaky abstraction here, but it feels like a place where dbt can help in a big way, and where we won't see the full value of Python models without some answer here.

Questions we need to answer:
- Is this "upload" step part of `dbt deps`? What if these reusable methods live in the root project, rather than a dbt package?
- How to support the transition from local &rarr; cloud development (if we want to preserve that workflow)?
- Those Python models should ideally be (a) pure functions with (b) thorough unit testing. Can dbt help here? Or should we encourage standard best practices Python development (`pytest` etc)?

@lostmygithubaccount @ChenyuLInx This feels like one of our highest-priority spikes during the v1.3 beta &rarr; final period.